File: 8. String to Integer (atoi)/8. String to Integer (atoi).ts
function myAtoi(s: string): number {
    let i = 0;
    const n = s.length;
    
    // Step 1: Ignore leading whitespace
    while (i < n && s[i] === ' ') {
        i++;
    }
    
    // Step 2: Check for sign
    let sign = 1;
    if (i < n && (s[i] === '-' || s[i] === '+')) {
        sign = s[i] === '-' ? -1 : 1;
        i++;
    }
    
    // Step 3: Convert digits
    let result = 0;
    while (i < n && s[i] >= '0' && s[i] <= '9') {
        const digit = parseInt(s[i]);
        // Check for overflow
        if (result > Math.floor((2**31 - 1) / 10) || (result === Math.floor((2**31 - 1) / 10) && digit > 7)) {
            return sign === 1 ? 2**31 - 1 : -(2**31);
        }
        result = result * 10 + digit;
        i++;
    }
    
    return sign * result;
};

File: 8. String to Integer (atoi)/8. String to Integer (atoi).go
func myAtoi(s string) int {
    i, sign, result := 0, 1, 0
    
    // Step 1: Ignore leading whitespace
    for i < len(s) && s[i] == ' ' {
        i++
    }
    
    // Step 2: Check for sign
    if i < len(s) && (s[i] == '-' || s[i] == '+') {
        if s[i] == '-' {
            sign = -1
        }
        i++
    }
    
    // Step 3: Convert digits
    for i < len(s) && s[i] >= '0' && s[i] <= '9' {
        digit := int(s[i] - '0')
        // Check for overflow
        if result > math.MaxInt32/10 || (result == math.MaxInt32/10 && digit > math.MaxInt32%10) {
            if sign == 1 {
                return math.MaxInt32
            }
            return math.MinInt32
        }
        result = result*10 + digit
        i++
    }
    
    return sign * result
}